daemon/policy.go

```diff
@@ -282,6 +311,35 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
 	return rev, nil
 }
 
+// ReactToRuleUpdates waits until wg is complete to do the following
+// * regenerate all endpoints in epsToRegen
+// * bump the policy revision of all endpoints not in epsToRegen, but which are
+//   in allEps, to revision rev.
+func (d *Daemon) ReactToRuleUpdates(wg *sync.WaitGroup, allEps []policy.IdentityConsumer, epsToRegen *policy.IDSet, rev uint64) {
+	// Wait until we have calculated which endpoints need to be selected
+	// across multiple goroutines.
+	wg.Wait()
+
+	epsToRegen.Mutex.RLock()
+	defer epsToRegen.Mutex.RUnlock()
+
+	// If an endpoint is not in the list of endpoints to regenerate, then the
+	// policy revision for the endpoint needs to be bumped so we can reflect that
+	// said endpoint realizes the state of the policy repository at revision rev.
+	for _, ep := range allEps {
+		if ep == nil {
+			continue
+		}
+		epID := ep.GetID16()
+		if _, ok := epsToRegen.IDs[epID]; !ok {
+			go ep.PolicyRevisionBumpEvent(rev)
+		}
+	}
+
+	// Regenerate all other endpoints.
+	endpointmanager.RegenerateEndpointSet(d, &endpoint.ExternalRegenerationMetadata{Reason: "policy rules added"}, epsToRegen.IDs)
```

Comment:
This function is called when policy rules are deleted, too. The reason is now lost with this refactor. The caller should probably pass it in.

Reply:
Oh, and be aware also of #7277 which adjust the structure of the `ExternalRegenerationMetadata`. Not sure if it will have a meaningful effect on calls like this just yet, but if it ends up going in first then it'd be good to review.